### PR TITLE
[NJT] Support Chunk backward for simple cases

### PR DESCRIPTION
**Pull request by YuqingJ**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #132193

### CODE REVIEW

File: aten/src/ATen/native/TensorShape.cpp [hunk 1]
```diff
@@ -4090,4 +4091,11 @@ int64_t dense_dim_default(const Tensor& self) {
   return self.dim();
 }
 
+Tensor chunk_backward(at::TensorList grads, const at::Tensor & input, int64_t chunks, int64_t dim) {
+    // Not implemented
+    // chunk_backward is only used by Nested Jagged Tensor.
+    TORCH_CHECK_NOT_IMPLEMENTED(input.is_nested() && input.layout() == c10::kJagged, "chunk_backward is not implemented explicitly");
```
Comment (davidberard98) [inline aten/src/ATen/native/TensorShape.cpp:4097]: nit: should this always assert? if it's a NJT then I would imagine it should directly go to the ops.py implementation, otherwise that seems unexpected

**Review by davidberard98 — COMMENTED** [review]
a few minor comments, otherwise lgtm! but leaving as unaccepted so this still shows up in the review queue for @soulitzer @jbschlosser, in case there's anything I missed

File: aten/src/ATen/native/TensorShape.cpp (hunk 1 above)
Comment (YuqingJ) [reply aten/src/ATen/native/TensorShape.cpp:4097]: This assert is for NST.

Comment (davidberard98) [reply aten/src/ATen/native/TensorShape.cpp:4097]: right, but even in the case where input _is_ NJT, it seems like we should still throw an exception? (which doesn't happen now I think)

File: tools/autograd/derivatives.yaml [hunk 2]
```diff
@@ -397,6 +397,14 @@
 - name: cholesky(Tensor self, bool upper=False) -> Tensor
   self: cholesky_backward(grad, upper, result)
 
+- name: chunk(Tensor(a -> *) self, int chunks, int dim=0) -> Tensor(a)[]
+  dispatch:
+    Default:
+      # the default case will use the CompositeImplicitAutograd
+      self: not_implemented("chunk")
+    AutogradNestedTensor:
+      self: chunk_backward(grads, self, chunks, dim)
```
[the remark this answers is not in the document]
Comment (YuqingJ) [reply tools/autograd/derivatives.yaml:406]: Does self.sym_sizes() contain the offsets too?

File: torch/nested/_internal/ops.py [hunk 3]
```diff
@@ -724,6 +724,53 @@ def chunk_default(func, *args, **kwargs):
...
+    for grad in grads:
+        if grad is None:
+            need_to_fill_grad = True
+            break
+
+    if need_to_fill_grad:
+        # Since torch.zeros([.., j0, ..]) is not supported, skipping this case for now
+        raise RuntimeError(
+            "NestedTensor chunk backward requires that all grads are not NONE"
+        )
```
[the remark this answers is not in the document]
Comment (soulitzer) [reply torch/nested/_internal/ops.py:754]: The issue is that saving just shape is not enough because the shape itself does not contain the raggedness information today. We'd be able to easily support this later when we store the offsets on shape as part of the factory function support stack, but it is tricky today.

File: aten/src/ATen/native/native_functions.yaml [hunk 4]
```diff
@@ -1466,6 +1466,12 @@
     CompositeImplicitAutograd: chunk
     NestedTensorCPU, NestedTensorCUDA: chunk_nested_tensor
 
+- func: chunk_backward(Tensor[] grads, Tensor input, int chunks, int dim) -> Tensor
+  variants: function
+  dispatch:
+    CompositeImplicitAutograd: chunk_backward
+    NestedTensorCPU, NestedTensorCUDA: chunk_backward
+
```
Comment (soulitzer) [inline aten/src/ATen/native/native_functions.yaml:1474]: If we only handle the all grads non-None case, it sounds like we can avoid this native function. We could just have a function in torch/csrc/autograd/FunctionsManual.cpp that loops through grads to verify non-None and then calls into cat.

Comment (YuqingJ) [reply aten/src/ATen/native/native_functions.yaml:1474]: Just added the support for the cases when some grad are none too.

Comment (soulitzer) [reply aten/src/ATen/native/native_functions.yaml:1474]: I think we can still do it without native functions. Since we are saving self, you could reapply chunk, then zeros_like the ones corresponding to None.

File: aten/src/ATen/native/TensorShape.cpp (hunk 1 above)
Comment (soulitzer) [reply aten/src/ATen/native/TensorShape.cpp]: nit: Maybe `TORCH_INTERNAL_ASSERT` instead? `TORCH_CHECK_NOT_IMPLEMENTED` implies that we'd want to eventually implement it.

File: aten/src/ATen/native/native_functions.yaml (hunk 4 above)
Comment (YuqingJ) [reply aten/src/ATen/native/native_functions.yaml:1474]: I see. If we want to support backward for chunk(dim=0) eventually, we will need to do it through native functions right?

Comment (soulitzer) [reply aten/src/ATen/native/native_functions.yaml:1474]: Actually the better way is to create a torch.zeros_like(self), chunk, copy_ each grad into the chunk outputs. We should do that in this PR today if possible since its more efficient. This would also avoid the issue of the cat(dim=0) producing a brand new offsets.

Comment (davidberard98) [reply aten/src/ATen/native/native_functions.yaml:1474]: @soulitzer , zeros_like is a good point.

But re: making a new op: in order to know where to copy_ the grads, don't you need to be able to see the offsets, which you'd need to do in ops.py?

Comment (soulitzer) [reply aten/src/ATen/native/native_functions.yaml:1474]: chunk(dim=0) should return a bunch of views, these views correspond one-to-one with the grad outputs you receive, s.t. the shape of the view is exactly the same as the shape of the grad. So all you need to do is `chunk(dim=0)[i].copy_(grad_outputs)`.

Comment (YuqingJ) [reply aten/src/ATen/native/native_functions.yaml:1474]: Thanks for the suggestion, it is definitely a cleaner approach. 
One question about the copy_: We don't have copy_ support for NJT yet, I can add it but I am wondering how we should check the offsets of the self and src. I think the offsets need to be the same, but not necessary the exact same tensor. But today we can only check if the offsets are the exact same tensor in NJT, which doesn't work for the chunk backend usage. 
This won't be an issue if we implement this in ops.py since we can just copy_ the values.

Comment (soulitzer) [reply aten/src/ATen/native/native_functions.yaml:1474]: That's a good point. I wonder if we could eventually have some kind of new copy_unsafe_ op, whose semantics are that of copy_ except we ignore the shapes (safe in this context because autograd has already validated the shapes).
We'd be able to reuse this when we later add dim=0 support to unbind_backward_nested_jagged which @jbschlosser has also already implemented with the replay view + copy_ approach.

Another option is to eventually define the union find merge operations as an aten ops so that we can access them in derivatives. But seems a little wasteful because as soon as we return from this backward op, we'll never see the offsets of the replayed view again.

File: test/functorch/test_vmap_registrations.py [hunk 5]
```diff
@@ -47,6 +47,7 @@
     "aten::chain_matmul",
     "aten::chalf",
     "aten::choose_qparams_optimized",
+    "aten::chunk_backward",
```
Comment (soulitzer) [inline test/functorch/test_vmap_registrations.py:50]: forgot to remove this?

File: torch/nested/_internal/ops.py [hunk 6]
```diff
@@ -465,6 +465,24 @@ def to_copy_default(func, *args, **kwargs):
...
+    _, new_kwargs = normalize_function(
+        func, args=args, kwargs=kwargs, normalize_to_only_use_kwargs=True
+    )
+    inp = new_kwargs.pop("input")
+    src = new_kwargs.pop("src")
+    check_ragged_dim_same(func, inp, "self", src, "src")
+    if inp._ragged_idx != src._ragged_idx:
+        raise RuntimeError(
+            "copy_ only supports tensors that are the same size for Nested Tensor"
+        )
```
Comment (soulitzer) [inline torch/nested/_internal/ops.py:481]: Can we just do `inp.shape == src.shape`?

File: test/test_nestedtensor.py [hunk 7]
```diff
@@ -3917,18 +3917,45 @@ def test_flatten_decomp(self, device):
...
+            5, 11, dtype=torch.float64, device=device, requires_grad=True
+        )
+        offsets = torch.tensor([0, 2, 3, 5], device=device)
+
+        def grad_test_func(values, offsets):
+            nt = torch.nested.nested_tensor_from_jagged(values, offsets)
+            chunks = nt.chunk(3, dim=-1)
+            return chunks[0].values().sum()
+
+        assert gradcheck(
```
Comment (soulitzer) [inline test/test_nestedtensor.py:3953]: nit: self.assertTrue

**Review by soulitzer — APPROVED** [review]
Thanks! Some minor comments

File: test/test_nestedtensor.py (hunk 7 above)
Comment (YuqingJ) [reply test/test_nestedtensor.py:3953]: It seems that the whole file is using assert gradcheck(..) or just gradcheck(..). I guess they are both fine?

Comment (soulitzer) [reply test/test_nestedtensor.py:3953]: Ah I didn't realize `assert gradcheck` was preexisting, and yeah just gradcheck is fine too. The gradcheck in testing is configured to error if it would've returned False.
